sklearn/utils/sparsefuncs_fast.pyx

```diff
@@ -280,24 +366,62 @@ def _incr_mean_variance_axis0(np.ndarray[floating, ndim=1] X_data,
     updated_var = np.zeros_like(new_mean, dtype=dtype)
 
     cdef unsigned long new_n
+    cdef np.ndarray[unsigned int, ndim=1] new_n_feat
     cdef unsigned long updated_n
+    cdef np.ndarray[unsigned int, ndim=1] updated_n_feat
     cdef floating last_over_new_n
 
     # Obtain new stats first
     new_n = n_samples
 
     if X_format == 'csr':
         # X is a CSR matrix
+        new_n_feat = n_samples_count_csr(X_data, shape, X_indices)
         new_mean, new_var = _csr_mean_variance_axis0(X_data, shape, X_indices)
     else:
         # X is a CSC matrix
+        new_n_feat = n_samples_count_csc(X_data, shape, X_indices, X_indptr)
         new_mean, new_var = _csc_mean_variance_axis0(X_data, shape, X_indices,
                                                      X_indptr)
+    new_n = new_n_feat[0]
 
     # First pass
-    if last_n == 0:
+    if last_n == 0 and (last_n_feat==0).all():
         return new_mean, new_var, new_n
     # Next passes
+
+    # Where each feature has different values and updated_n_feat is a vector
+    elif last_n==0 and (last_n_feat!=0).any():
+        updated_n_feat = last_n_feat + new_n_feat
+
+        for i in xrange(n_features):
+            if updated_n_feat[i] == 0:
+                continue
+            if new_n_feat[i] == 0:
+                updated_mean[i] = last_mean[i]
+                updated_var[i] = last_var[i]
+                continue
+            last_over_new_n = last_n_feat[i] * 1.0 / new_n_feat[i]
+            # Unnormalized old stats
+            last_mean[i] *= last_n_feat[i]
+            last_var[i] *= last_n_feat[i]
+
+            # Unnormalized new stats
+            new_mean[i] *= new_n_feat[i]
+            new_var[i] *= new_n_feat[i]
+
+            # Update stats
+            updated_var[i] = (last_var[i] + new_var[i] +
+                              last_over_new_n / updated_n_feat[i] *
+                              (last_mean[i] / last_over_new_n -
+                               new_mean[i]) ** 2)
+
+            updated_mean[i] = (last_mean[i] + new_mean[i]) / updated_n_feat[i]
+            updated_var[i] = updated_var[i] / updated_n_feat[i]
+
+        return updated_mean, updated_var, updated_n_feat
+
+    # Where updated_n is a scaler
     else:
         updated_n = last_n + new_n
```

Comment:
I am proposing to simplify the code in this case using vectorized numpy functions in #10615. You should use vectorized functions for the above case (or even share the implementation in both cases) too.

Reply:
i agree, there are a lot places where i should use vectorized functions. Dealing with Sparse Matrix and incremental_mean_and_var was getting messed up with various kinds of errors. So I created a separate pull request to check if *incremental_mean_and_var** works without failing in numerical stability.

The basic idea/design is:
* if there is a NaN then avoid/ignore it and calculate mean and var (using the formula present in the code)

While doing this various cases comes up like:
- if the entire input matrix is NaN, then new_sample_count will be zero.
Consider that last_sample_count is also Zero then Updated_sample_count will also be zero.

Cases like this was not possible before. This kind of situations kind of lead me to use For loops a bit too much, but I shall change them by using vector functions.

In my last commit, I am also getting error in numerical stability hence the separate pull request to deal with 1 problem at a time and see it they work perfectly.

I am planning to deal with **incremental_mean_and_var** and **Sparse Matrix** separately and then combine them to get the final solution